### PR TITLE
refactor(experimental): add `getFeeForMessage` API method

### PR DESCRIPTION
**Pull request by buffalojoec**
This PR adds the `getFeeForMessage` RPC method to the new experimental Web3 JS's arsenal.

Ref #1449

### CODE REVIEW

File: packages/rpc-core/src/rpc-methods/getFeeForMessage.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+/** Fee corresponding to the message at the specified blockhash */
+type GetFeeForMessageApiResponse = RpcResponse<U64UnsafeBeyond2Pow53Minus1 | null>;
+
+export interface GetFeeForMessageApi {
+    /**
+     * Returns the fee the network will charge for a particular Message
+     */
+    getFeeForMessage(
+        /** Base-64 encoded message */
+        message: string, // TODO: May not be string
```
Comment (mcintyre94) [inline packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:12]: `Base64EncodedBytes`?

**Review by mcintyre94 — APPROVED** [review]
Looks good!

File: packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    });
+
+    (['confirmed', 'finalized', 'processed'] as Commitment[]).forEach(commitment => {
+        describe(`when called with \`${commitment}\` commitment`, () => {
+            if (commitment === 'finalized') {
+                it.todo(
+                    'returns the result as a bigint (test broken; see https://discord.com/channels/428295358100013066/560496939779620864/1132048104728825926)'
+                );
+                return;
+            }
```
Comment (mcintyre94) [inline packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts:74]: Hmm is this a timing issue that could affect `confirmed` too if the test runs early enough?

File: packages/rpc-core/src/rpc-methods/getFeeForMessage.ts (hunk 1 above)
Comment (buffalojoec) [reply packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:12]: I think that makes sense, is that de facto for tx message?

File: packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts (hunk 2 above)
Comment (buffalojoec) [reply packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts:74]: I'm not sure on confirmed. I plucked this from https://github.com/solana-labs/solana-web3.js/blob/master/packages/rpc-core/src/rpc-methods/__tests__/send-transaction-test.ts#L115-L120

It's definitely related to the fact that it takes ~17s for the test validator to finalize slot 0, but `confirmed` seems to be almost immediate so maybe that's not a concern? I don't see `confirmed` omitted anywhere else in our test suites

File: packages/rpc-core/src/rpc-methods/getFeeForMessage.ts (hunk 1 above)
Comment (mcintyre94) [reply packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:12]: I want to say yes but there's probably somewhere we use some other encoding for some reason 😅 In this case though the RPC docs are clear that it's base64 and that's exactly what this type is for!

Comment (buffalojoec) [reply packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:12]: Got it, going with that then!

File: packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            describe('when called with a recent blockhash', () => {
+                it('returns the result as a bigint', async () => {
+                    expect.assertions(1);
+                    const latestBlockhash = await rpc.getLatestBlockhash().send();
+                    const message = getMockTransactionMessage(latestBlockhash.value.blockhash);
+                    console.log('Printing mock message', message);
```
Comment (steveluscher) [inline packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts:80]: 👀

File: packages/rpc-core/src/rpc-methods/getFeeForMessage.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+/** Fee corresponding to the message at the specified blockhash */
+type GetFeeForMessageApiResponse = RpcResponse<U64UnsafeBeyond2Pow53Minus1 | null>;
+
+export interface GetFeeForMessageApi {
+    /**
+     * Returns the fee the network will charge for a particular Message
+     */
+    getFeeForMessage(
+        /** Base-64 encoded message */
+        message: Base64EncodedBytes,
```
Comment (steveluscher) [inline packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:12]: We should really create an opaque type here to represent ‘a bunch of bytes that are definitely a serialized message.’

1. Create such an opaque type
2. Update `packages/transaction/src/serializers/message.ts` to return that type from `serialize`
3. Make this function only accept that type

File: packages/rpc-core/src/rpc-methods/getFeeForMessage.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { Base64EncodedBytes, Commitment, RpcResponse, Slot, U64UnsafeBeyond2Pow53Minus1 } from './common';
+
+/** Fee corresponding to the message at the specified blockhash */
+type GetFeeForMessageApiResponse = RpcResponse<U64UnsafeBeyond2Pow53Minus1 | null>;
```
Comment (steveluscher) [inline packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:4]: The RPC does not wrap this in a `RpcResponse` wrapper.

```diff
- type GetFeeForMessageApiResponse = RpcResponse<U64UnsafeBeyond2Pow53Minus1 | null>;
+ type GetFeeForMessageApiResponse = U64UnsafeBeyond2Pow53Minus1 | null;
```

Comment (buffalojoec) [reply packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:4]: I believe it does mate

https://docs.solana.com/api/http#getfeeformessage
https://github.com/solana-labs/solana/blob/67d6d688ccf10d09a012559a43f524aa35e87919/rpc/src/rpc.rs#L3996
https://github.com/solana-labs/solana/blob/67d6d688ccf10d09a012559a43f524aa35e87919/rpc/src/rpc.rs#L120

File: packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+// See scripts/fixtures/send-transaction-fee-payer.json
+const MOCK_PUBLIC_KEY_BYTES = // DRtXHDgC312wpNdNCSb8vCoXDcofCJcPHdAw4VkJ8L9i
+    // prettier-ignore
+    new Uint8Array([
+        0xb8, 0xac, 0x70, 0x4f, 0xaf, 0xc7, 0xa5, 0xfc, 0x8c, 0x5d, 0x1f, 0x0a, 0xc8, 0xcf, 0xaa, 0xe0,
+        0x42, 0xfa, 0x3b, 0xb8, 0x25, 0xf0, 0xec, 0xfc, 0xe2, 0x27, 0x4d, 0x7d, 0xad, 0xad, 0x51, 0x2d,
+    ]);
+
+function getMockTransactionMessage(blockhash: Blockhash) {
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts:19]: I think the hard part is hard-coding a valid blockhash for the recent blockhash.
Even in the test for `sendTransaction` that's what we're doing. 

https://github.com/solana-labs/solana-web3.js/blob/dc614aff8d85929e95452e4af0e4f5d3310ee82e/packages/rpc-core/src/rpc-methods/__tests__/send-transaction-test.ts#L123-L131

A solution for blockhash might be to somehow grab it from the validator logs, then we get around the _retrieval_ of a valid blockhash, but if that blockhash isn't deterministic within our test suite then I don't know how we can hard-code the serialized value of it.

Blockhashes should be _kind of_ deterministic with a local validator right? I mean, if you know precisely the slots and transactions you can maybe reproduce it, but I'm not sure on the timing of all the test suites firing at once.

File: packages/rpc-core/src/rpc-methods/getFeeForMessage.ts (hunk 5 above)
Comment (steveluscher) [reply packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:4]: Oh, oops. I was reading just the docs, which are wrong. They should mention that an RpcResponse object is returned.

File: packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (steveluscher) [reply packages/rpc-core/src/rpc-methods/__tests__/get-fee-for-message-test.ts:19]: Oh crap. I wasn’t thinking of the fact that the fee schedule is tied to a moment in time, ie. a blockhash.

Carry on!

File: packages/rpc-core/src/rpc-methods/getFeeForMessage.ts (hunk 5 above)
Comment (buffalojoec) [reply packages/rpc-core/src/rpc-methods/getFeeForMessage.ts:4]: Yeah looks like it's in their payload just not the description like the others

**Pull request merged**